clippy_lints/src/single_field_patterns.rs

```diff
@@ -0,0 +1,264 @@
+///   where None represents the absence of a lint
+#[derive(Debug, Clone, Copy)]
+enum SingleField {
+    Id { id: Ident, pattern: Span },
+    Unused, // The name "SingleField" is a lie but idk what's better. "AtMostOneField"?
+}
+
+impl PartialEq for SingleField {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (SingleField::Id { id: id1, .. }, SingleField::Id { id: id2, .. }) => id1 == id2,
+            (SingleField::Unused, SingleField::Unused) => true,
+            _ => false,
+        }
+    }
+}
+
+fn get_sf<'a>(mut iter: impl Iterator<Item = (Ident, &'a Pat<'a>)>) -> Option<SingleField> {
+    let one = iter.by_ref().find(|(_, pat)| !matches!(pat.kind, PatKind::Wild));
+    match one {
+        Some((id, pat)) => {
+            if pat.span.from_expansion() {
+                return None;
+            }
+            if iter.all(|(_, other)| matches!(other.kind, PatKind::Wild)) {
+                Some(SingleField::Id { id, pattern: pat.span })
+            } else {
+                None
+            }
+        },
+        None => Some(SingleField::Unused),
+    }
+}
+
+fn struct_sf(pat: &PatKind<'_>) -> Option<SingleField> {
+    match *pat {
+        PatKind::Struct(_, pats, _) => get_sf(pats.iter().map(|field| (field.ident, field.pat))),
+        _ => None,
+    }
+}
+
+/// This handles recursive patterns and flattens them out lazily
+/// e.g. 1 | (2 | 9) | 3..5
+struct FlatPatterns<'hir, I>
+where
+    I: Iterator<Item = &'hir Pat<'hir>>,
+{
+    patterns: I,
+    stack: Vec<&'hir Pat<'hir>>,
+}
+
+impl<I: Iterator<Item = &'hir Pat<'hir>>> FlatPatterns<'hir, I> {
+    fn new(patterns: I) -> Self {
+        Self {
+            patterns,
+            stack: Vec::new(),
+        }
+    }
+}
+
+impl<I: Iterator<Item = &'hir Pat<'hir>>> Iterator for FlatPatterns<'hir, I> {
+    type Item = &'hir Pat<'hir>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stack.is_empty() {
+            if let Some(pat) = self.patterns.next() {
+                self.stack.push(pat);
+            } else {
+                return None;
+            }
+        }
+        while let Some(pat) = self.stack.pop() {
+            match pat.kind {
+                PatKind::Or(pats) => self.stack.extend(pats),
+                _ => return Some(pat),
+            }
+        }
+        unreachable!("Or always has 2 patterns, so one of the prior returns must return");
+    }
+}
+
+fn find_sf_lint<'hir>(
+    cx: &LateContext<'_>,
+    patterns: impl Iterator<Item = &'hir Pat<'hir>>,
+    leaf_sf: &impl Fn(&PatKind<'hir>) -> Option<SingleField>,
+) -> Option<(SingleField, Vec<(Span, String)>)> {
+    let fields = FlatPatterns::new(patterns).map(|p| {
+        (
+            p.span,
+            if matches!(p.kind, PatKind::Wild) {
+                Some(SingleField::Unused)
+            } else {
+                leaf_sf(&p.kind)
+            },
+        )
+    });
```

Comment:
```suggestion
    let fields = FlatPatterns::new(patterns).filter_map(|p| {
        if matches!(p.kind, PatKind::Wild) {
            Some(SingleField::Unused)
        } else {
            leaf_sf(&p.kind)
        }.map(|sf| (p.span, sf))
    });
```

With that you don't have to check for `Some(sf) = sf` in the for loop.

Reply:
filter_map has slightly different semantics. leaf_sf (which I've since inlined), returns None on PatKind::Binding - and we shouldn't lint one of those